daemon/worker.c

```diff
@@ -1029,15 +1045,124 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
 			comm_point_drop_reply(repinfo);
 			return 0; /* discard this */
 		}
-		sldns_buffer_set_limit(c->buffer, LDNS_HEADER_SIZE);
-		sldns_buffer_write_at(c->buffer, 4, 
-			(uint8_t*)"\0\0\0\0\0\0\0\0", 8);
+		log_assert(LDNS_QDCOUNT(sldns_buffer_begin(c->buffer)) == 1);
+
+		sldns_buffer_skip(c->buffer, LDNS_HEADER_SIZE); /* skip header */
+
+		if (!query_dname_len(c->buffer)) {
+			LDNS_QR_SET(sldns_buffer_begin(c->buffer));
```

Comment:
Since the qname had a parse error, we should not return it so we return at least a **valid DNS error**.

```suggestion
			LDNS_QDCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_ANCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_NSCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_ARCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_QR_SET(sldns_buffer_begin(c->buffer));
```